File: practices/interpreterbook/src/02/token/token.go
// Go言語で作るインタプリタ 1
package token

type TokenType string

type Token struct {
  Type    TokenType
  Literal string
}

var keywords = map[string]TokenType{
  "fn":     FUNCTION,
  "let":    LET
  "true":   TRUE,
  "false":  FALSE,
  "if":     IF,
  "else":   ELSE,
  "return": RETURN,
}

// identがキーワードの場合: TokenTypeトークン
// そうでない場合: IDENTトークン
func LookupIdent(ident string) TokenType {
  if tok, ok := keywords[ident]; ok {
    return tok
  }
  return IDENT
}

const (
  ILLEGAL = "ILLEGAL"
  EOF     = "EOF"

  // 識別子・リテラル
  IDENT = "IDENT"
  INT   = "INT"

  // 演算子
  BANG     = "!"
  ASSIGN   = "="
  PLUS     = "+"
  MINUS    = "-"
  ASTERISK = "*"
  SLASH    = "/"

  LT = "<"
  GT = ">"

  EQ     = "=="
  NOT_EQ = "!="

  // デリミタ
  COMMA     = ","
  SEMICOLON = ";"

  LPARAN = "("
  RPARAN = ")"
  LBRACE = "{"
  RBRACE = "}"

  // キーワード
  FUNCTION = "FUNCTION"
  LET      = "LET"
)

File: practices/interpreterbook/ruby/lib/token.rb
class Token
  ILLEGAL   = "ILLEGAL"
  EOF       = "EOF"
  IDENT     = "IDENT"
  INT       = "INT"
  BANG      = "!"
  ASSIGN    = "="
  PLUS      = "+"
  MINUS     = "-"
  ASTERISK  = "*"
  SLASH     = "/"
  LT        = "<"
  GT        = ">"
  EQ        = "=="
  NOT_EQ    = "!="
  COMMA     = ","
  SEMICOLON = ";"
  LPAREN    = "("
  RPAREN    = ")"
  LBRACE    = "{"
  RBRACE    = "}"
  FUNCTION  = "FUNCTION"
  LET       = "LET"
  TRUE      = "TRUE"
  FALSE     = "FALSE"
  IF        = "IF"
  ELSE      = "ELSE"
  RETURN    = "RETURN"


  class << self
    def lookup_identifier(ident)
      keywords.fetch(ident, IDENT)
    end

    private

    def keywords
      {
        "fn"     => FUNCTION,
        "let"    => LET,
        "true"   => TRUE,
        "false"  => FALSE,
        "if"     => IF,
        "else"   => ELSE,
        "return" => RETURN,
      }
    end
  end

  attr_accessor :type, :literal

  def initialize(type: nil, literal: nil)
    @type = type
    @literal = literal
  end
end
